### PR TITLE
adds base64 decoding (fixes #5656)

### PR DESCRIPTION
**Pull request by amartgon**
fixes #5656

### CODE REVIEW

**Comment by ScottPJones** [comment]
Would there be any interest in adding RFC4648 compliant "URL and Filename safe" base64 encoding to this (i.e. base64url)?  This is a fairly common use case...
It is only using a different array for the encoding... (i.e. `-_` instead of `+/`).
I had to write my own, but it would be nice if it were part of Julia instead.

**Comment by tkelman** [comment]
Not sure base64 encoding is really fundamental enough to need in base at all. If Codecs.jl has some existing base64 encoding functionality, that would be a good place to extend it with additional variants.

**Comment by ScottPJones** [comment]
@tkelman Frankly, with all the other stuff in Base, (multimedia.jl?), and with base64 already in Base, that's where I thought it would have to be...  I'll look at Codecs.jl [but to be honest... how many people looking for base64 encoding are going to think about codecs? ;-) ]  Thanks!

**Comment by tkelman** [comment]
Package discoverability is a problem. There are multiple issues open on that.

base64 is in base because no one has tried to remove `stringmime` which appears to be the only place in base that it gets used. The MIME types are used by IJulia and a few other places for rich output display, but I imagine they could be adapted to get their base64 encoding from an external package if they needed to. #5155 is a very slow and arduous issue to make progress on, since removing code from base is harder to do in a painless way than adding code to base. But going further in the direction of "anything anyone ever found useful enough to implement in a PR to JuliaLang/julia" is probably not ideal for long-term maintainability.

**Comment by ScottPJones** [comment]
@tkelman I agree totally... I think there's way too much stuff there as it is! ;-)  I'm *totally* fine with adding RFC4648 base64url to Codecs.jl, if people think it would be useful (besides me, who's already using it in my own ocde), and that is the right place.

**Comment by timholy** [comment]
Agreed with the package idea.

File: base/base64.jl [hunk 1]
```diff
@@ -134,4 +165,48 @@ base64(x...) = base64(write, x...)
...
+    return length(b.cache) == 0 && eof(b.io)
+end
+
+function close(b::Base64DecodePipe)
+end
+
+# Decodes a base64-encoded string
+function base64decode(s)
+    b = IOBuffer(s)
+    decoded = readall(Base64DecodePipe(b))
```
Comment (stevengj) [inline base/base64.jl]: I think this is wrong.  It should be `readbytes`, not `readall`.  The Base64-encoded data need not produce valid UTF-8 string.

Comment (ScottPJones) [inline base/base64.jl]: Yes, that's definitely wrong...  esp. since most uses of base64 encoding are to store binary data, not Unicode text...

**Pull request merged**